File: week_1_sprint/two_pointers/python/container_with_most_water.py
# https://leetcode.com/problems/container-with-most-water/description/

from typing import List

class Solution:
    def maxArea(self, height: List[int]) -> int:
        # 2 pointers
        # Time Complexity O(n)
        # Space Complexity O(1)

        max_area: int = 0
        left: int = 0
        right: int = len(height) - 1

        while (left < right):
            width: int = right - left
            area: int = width * min(height[left], height[right])

            max_area = max(max_area, area)
            if (height[left] <= height[right]):
                left += 1
            else:
                right -= 1
        return max_area

File: leetcode/two-pointer-sliding-window/container-with-most-water.cpp
// https://leetcode.com/problems/container-with-most-water/

#include <vector>
using namespace std;

class Solution {
public:
    int maxArea(vector<int>& height) {
        return my_two_pointers(height);
    }

    // My Soln
    // Two Pointers 
    // Time Complexity O(n)
    // Space Complexity O(1)
    int my_two_pointers(const vector<int>& height) {
        int low = 0;
        int high = height.size() - 1;

        int ans = -1;
        while (low < high) {
            ans = max(
                ans, 
                (high - low) * min(height[high], height[low])
            );
            if (height[high] < height[low]) high--;
            else low++;
        }
        return ans;
    }
};
